### PR TITLE
Sccache, needs, CLI update

### PR DESCRIPTION
**Pull request by TriplEight**
- `sccache` now uses redis
- `needs` make jobs run right after a dependent job is finished, no more waiting until stage is done.
- `cargo-contract` will be updated in 2 example jobs if git version is newer than one installed in docker image. It uses the newest `cargo install` [feature](https://github.com/rust-lang/cargo/pull/7560).

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -129,9 +125,7 @@ fmt:
   stage:                           workspace
   <<:                              *docker-env
   script:
-    - for crate in ${ALL_CRATES}; do
-        cargo fmt --verbose -- --check;
-      done
+    - cargo fmt --verbose -- --check
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Missing `--all` flag if you want to go without the loop.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -98,28 +104,18 @@ build-wasm:
...
 clippy-std:
   stage:                           workspace
   <<:                              *docker-env
-  dependencies:
-    - check-std
   script:
-    - for crate in ${ALL_CRATES}; do
-        cargo clippy --verbose --all-features -- -D warnings;
-      done
+    - cargo clippy --verbose --all-features -- -D warnings
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Missing the `--all` flag if you want to go without the loop ... however, I really do not recommend doing so. Why have you changed it back?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -98,28 +104,18 @@ build-wasm:
 test:
   stage:                           workspace
   <<:                              *docker-env
-  dependencies:
-    - check-std
   script:
-    - for crate in ${ALL_CRATES}; do
-        cargo test --verbose --all-features --release;
-      done
+    - cargo test --verbose --all-features --release
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Missing the `--all` flag if you want to go without the loop ... however, I really do not recommend doing so. Why have you changed it back?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -73,23 +72,30 @@ check-wasm:
...
 build-std:
   stage:                           workspace
   <<:                              *docker-env
-  dependencies:
-    - check-std
   script:
-    - for crate in ${ALL_CRATES}; do
-        cargo build --verbose --all-features --release;
-      done
+    - cargo build --verbose --all-features --release
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Missing the `--all` flag if you want to go without the loop ... however, I really do not recommend doing so. Why have you changed it back?

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -61,9 +62,7 @@ check-std:
   stage:                           check
   <<:                              *docker-env
   script:
-    - for crate in ${ALL_CRATES}; do
-        cargo check --verbose --all-features;
-      done
+    - cargo check --verbose --all-features;
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Missing the `--all` flag if you want to go without the loop ... however, I really do not recommend doing so. Why have you changed it back?

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -171,19 +165,29 @@ examples-clippy-wasm:
...
+    - build-cli
   script:
+    - ls -la
+    - ls ./target/release/
+    - ./target/release/cargo-contract -V
     - for example in examples/lang2/*; do
-        pushd $example &&
-        cargo contract build &&
+        pushd $example && pwd && echo ${CI_BUILDS_DIR}
+        ${CI_PROJECT_DIR}/target/release/cargo-contract build &&
```
Comment (Robbepop) [inline .gitlab-ci.yml]: If you properly installed `cargo-contract` you normally don't need to access it that way.

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -12,6 +12,7 @@ stages:
 
 variables:
   GIT_STRATEGY:                    fetch
+  GIT_DEPTH:                       "3"
```
[the remark this answers is not in the document]
Comment (TriplEight) [reply .gitlab-ci.yml]: no any reason, let's change it later if needed

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -180,10 +180,11 @@ examples-contract-build:
   needs:
     - build-cli
   script:
-    - ./target/release/cargo-contract -V
+    - cp target/release/cargo-contract ~/.cargo/bin/
```
[the remark this answers is not in the document]
Comment (TriplEight) [reply .gitlab-ci.yml]: will do if there will be more than 2 dupes

File: .gitlab-ci.yml [hunk 9]
```diff
@@ -72,12 +75,35 @@ check-wasm:
         cargo check --verbose --no-default-features --target wasm32-unknown-unknown --manifest-path ${crate}/Cargo.toml;
       done
 
+build-cli:
+  stage:                           check
+  <<:                              *docker-env
+  script:
+    # - cargo build --manifest-path cli/Cargo.toml --release
+    - cargo install --git https://github.com/paritytech/cargo-contract --force
```
[the remark this answers is not in the document]
Comment (TriplEight) [reply .gitlab-ci.yml]: will fix it as soon as you'll switch to stable contract version

File: .gitlab-ci.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (Robbepop) [reply .gitlab-ci.yml]: ok

File: .gitlab-ci.yml [hunk 10]
```diff
@@ -12,8 +12,8 @@ stages:
 
 variables:
   GIT_STRATEGY:                    fetch
+  GIT_DEPTH:                       "3"
   CARGO_HOME:                      "/ci-cache/${CI_PROJECT_NAME}/cargo/${CI_JOB_NAME}"
-  SCCACHE_DIR:                     "/ci-cache/${CI_PROJECT_NAME}/sccache"
```
Comment (Robbepop) [inline .gitlab-ci.yml]: no more sccache?

File: .gitlab-ci.yml [hunk 11]
```diff
@@ -44,6 +44,8 @@ variables:
     - schedules
     - web
     - /^[0-9]+$/                   # PRs
+  dependencies:                    []
```
Comment (Robbepop) [inline .gitlab-ci.yml]: what does this do? Please ELI5

File: .gitlab-ci.yml [hunk 12]
```diff
@@ -72,12 +74,27 @@ check-wasm:
...
+  stage:                           check
+  <<:                              *docker-env
+  variables:
+    GIT_STRATEGY:                    none
+  script:
+    - cargo install --git https://github.com/paritytech/cargo-contract --root ./artifacts
+    - ./artifacts/bin/cargo-contract -V
+  artifacts:
+    name:                          cli_${CI_COMMIT_SHORT_SHA}
+    expire_in:                     24 hrs
```
Comment (Robbepop) [inline .gitlab-ci.yml]: So it will only be build once every 24 hours? If so this is a good start until we support it on crates.io and make it usable from `cargo-update` crate.

File: .gitlab-ci.yml [hunk 13]
```diff
@@ -168,7 +185,13 @@ examples-clippy-wasm:
 examples-contract-build:
   stage:                           examples
   <<:                              *docker-env
+  dependencies:
+    - build-cli
+  needs:
+    - build-cli
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Why does it need both? Isn't `dependencies: build-cli` enough here?

File: .gitlab-ci.yml (hunk 10 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: It went Redis.

File: .gitlab-ci.yml (hunk 11 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: The default is every job downloads artifacts.
Empty array means it's not needed.
I put this in the template to skip artifacts upload by default.
And it's overriden in two jobs at the bottom.

File: .gitlab-ci.yml (hunk 12 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: Since every pipeline rebuilds it, you will have this artifact in every pipeline. 
So if it expires for the given pipeline, to restart the dependant job, you'll have to restart the builder job.

File: .gitlab-ci.yml (hunk 13 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: I'll check later, but `dependencies` are set for uploading artifacts from that job. `needs` is to start right after that job succeds.

Comment (Robbepop) [reply .gitlab-ci.yml]: :ok_hand:
